### PR TITLE
Add Param and methods to Spark.ML

### PR DESCRIPTION
**Pull request by GoEddie**
This adds the Param class and the appropriate methods on FeatureBase to be shared by all the Features that implement `https://spark.apache.org/docs/1.5.1/api/java/org/apache/spark/ml/param/Params.html`.

The scala class is `org.apache.spark.ml.param.Param` but the .NET naming standard is `Spark.ML.Param.Param` so defining it in C# means using `Param.Param` - not sure if that is acceptable or if it is worth moving it up a level in the namespace?

### CODE REVIEW

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        
+        JvmObjectReference IJvmObjectReferenceProvider.Reference => _jvmObject;
+        
+        /// <summary>
+        /// The description of what the <see cref="Param"/> does and how it works including any
+        /// defaults and the current value
+        /// </summary>
+        /// <returns>A description of how the <see cref="Param"/> works</returns>
+        public string Doc() => (string)_jvmObject.Invoke("doc");
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:65]: Should these be properties? I see the following on the Scala side:
```scala
class Param[T](val parent: String, val name: String, val doc: String, val isValid: T => Boolean)
```

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                s_ParamClassName, parent.Uid(), name, doc);
+        }
+        
+        public Param(string parent, string name, string doc)
+        {
+            _jvmObject = 
+                SparkEnvironment.JvmBridge.CallConstructor(s_ParamClassName, parent, name, doc);
+        }
+        
+        public Param(Identifiable parent, string name, string doc, bool isValid)
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:39]: Is this correct? `isValid`'s type is `T => Boolean`.

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs (hunk 1 above)
Comment (GoEddie) [reply src/csharp/Microsoft.Spark/ML/Param/Param.cs:65]: yes, I was looking at a really old version of the docs where it used to be a method, will change to a property

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs (hunk 2 above)
Comment (GoEddie) [reply src/csharp/Microsoft.Spark/ML/Param/Param.cs:39]: no, it was wrong, will change it (or remove it)

Comment (GoEddie) [reply src/csharp/Microsoft.Spark/ML/Param/Param.cs]: I removed it, to implement it requires quite a lot and it isn't something that I have seen really used and isn't something I need.

**Comment by imback82** [comment]
@suhsteve / @Niharikadutta Can you help with review?

File: src/csharp/Microsoft.Spark/ML/Feature/FeatureBase.cs [hunk 3]
```diff
@@ -56,7 +56,49 @@ internal FeatureBase(JvmObjectReference jvmObject)
         public T Save(string path) => 
             WrapAsType((JvmObjectReference)_jvmObject.Invoke("save", path));
 
-        private T WrapAsType(JvmObjectReference reference)
+        /// <summary>
+        /// Clears any value that was previously set for this <see cref="Param"/>. The value is
+        /// reset to the default value.
+        /// </summary>
+        /// <param name="param"></param>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/FeatureBase.cs:63]: Can you fill this?

File: src/csharp/Microsoft.Spark/ML/Feature/FeatureBase.cs [hunk 4]
```diff
@@ -56,7 +56,49 @@ internal FeatureBase(JvmObjectReference jvmObject)
...
+        /// <returns>Object reference that was used to clear the <see cref="Param"/></returns>
+        public T Clear(Param.Param param) => 
+            WrapAsType((JvmObjectReference)_jvmObject.Invoke("clear", param));
+
+        /// <summary>
+        /// Returns a description of how a specific <see cref="Param"/> works and is currently set.
+        /// </summary>
+        /// <param name="param"></param>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/FeatureBase.cs:71]: ditto

File: src/csharp/Microsoft.Spark/ML/Feature/FeatureBase.cs [hunk 5]
```diff
@@ -56,7 +56,49 @@ internal FeatureBase(JvmObjectReference jvmObject)
...
+        /// work and how they are currently set.
+        /// </summary>
+        /// <returns>Description of all the applicable <see cref="Param"/>'s</returns>
+        public string ExplainParams() => (string)_jvmObject.Invoke("explainParams");
+
+        /// <summary>
+        /// Retrieves a <see cref="Param"/> so that it can be used to set the value of the
+        /// <see cref="Param"/> on the object.
+        /// </summary>
+        /// <param name="paramName"></param>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/FeatureBase.cs:87]: ditto

File: src/csharp/Microsoft.Spark/ML/Feature/FeatureBase.cs [hunk 6]
```diff
@@ -56,7 +56,49 @@ internal FeatureBase(JvmObjectReference jvmObject)
...
+        /// <summary>
+        /// Returns a description of how a specific <see cref="Param"/> works and is currently set.
+        /// </summary>
+        /// <param name="param"></param>
+        /// <returns>Description of the <see cref="Param"/></returns>
+        public string ExplainParam(Param.Param param) => 
+            (string)_jvmObject.Invoke("explainParam", param);
+
+        /// <summary>
+        /// Returns a description of how all of the <see cref="Param"/> that apply to this object
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/FeatureBase.cs:77]: nit: `<see cref="Param"/>s`? Also no `'` below after `<see cref="Param"/>`?

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/BucketizerTests.cs [hunk 7]
```diff
@@ -58,6 +59,19 @@ public void TestBucketizer()
...
+            Param handleInvalidParam = bucketizer.GetParam("handleInvalid");
+            Assert.NotEmpty(handleInvalidParam.Doc);
+            Assert.NotEmpty(handleInvalidParam.Name);
+            Assert.Equal(handleInvalidParam.Parent, bucketizer.Uid());
+
+            Assert.NotEmpty(bucketizer.ExplainParam(handleInvalidParam));
+            bucketizer.Set(handleInvalidParam, "keep");
+            Assert.Equal("keep", bucketizer.GetHandleInvalid());
+             
+            Assert.Equal("error", bucketizer.Clear(handleInvalidParam).GetHandleInvalid());
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/BucketizerTests.cs:74]: I think we could add a suite for testing `Param` class on its own?

File: src/csharp/Microsoft.Spark/Microsoft.Spark.csproj [hunk 8]
```diff
@@ -38,10 +38,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <Content Include="..\..\scala\microsoft-spark-*\target\microsoft-spark-*.jar"
-             Link="jars\%(Filename)%(Extension)"
-             Pack="true"
-             PackagePath="jars\%(Filename)%(Extension)" />
+    <Content Include="..\..\scala\microsoft-spark-*\target\microsoft-spark-*.jar" Link="jars\%(Filename)%(Extension)" Pack="true" PackagePath="jars\%(Filename)%(Extension)" />
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Microsoft.Spark.csproj:41]: Can we revert this change?

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+namespace Microsoft.Spark.ML.Feature.Param
+{
+    /// <summary>
+    /// A <see cref="Param"/> with self-contained documentation and optionally default value.
+    ///
+    /// A <see cref="Param"/> references an individual parameter that includes documentation, the
+    /// name of the parameter and optionally a default value. Params can either be set using the
+    /// generic <see cref="Param"/> methods or by using explicit methods. For example
+    /// <see cref="Bucketizer"/> has SetHandleInvalid or you can call GetParam("handleInvalid")
+    /// and then <see cref="Bucketizer"/>.Set using the <see cref="Param"/> and the value you want
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:18]:
```suggestion
    /// and then <see cref="Bucketizer"/>. Set using the <see cref="Param"/> and the value you want
```

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+namespace Microsoft.Spark.ML.Feature.Param
+{
+    /// <summary>
+    /// A <see cref="Param"/> with self-contained documentation and optionally default value.
+    ///
+    /// A <see cref="Param"/> references an individual parameter that includes documentation, the
+    /// name of the parameter and optionally a default value. Params can either be set using the
+    /// generic <see cref="Param"/> methods or by using explicit methods. For example
+    /// <see cref="Bucketizer"/> has SetHandleInvalid or you can call GetParam("handleInvalid")
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:17]: Fix line length if you add this change.
```suggestion
    /// <see cref="Bucketizer"/> has <c>SetHandleInvalid</c> or you can call <c>GetParam("handleInvalid")</c>
```

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        private static readonly string s_ParamClassName = 
+            "org.apache.spark.ml.param.Param";
+        
+        private readonly JvmObjectReference _jvmObject;
+        
+        public Param(Identifiable parent, string name, string doc)
+        {
+            _jvmObject = SparkEnvironment.JvmBridge.CallConstructor(
+                s_ParamClassName, parent.Uid(), name, doc);
+        }
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:32]:
```suggestion
        public Param(Identifiable parent, string name, string doc)
            : this(SparkEnvironment.JvmBridge.CallConstructor(
                s_ParamClassName, parent.Uid(), name, doc)
        {
        }
```

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        
+        public Param(string parent, string name, string doc)
+        {
+            _jvmObject = 
+                SparkEnvironment.JvmBridge.CallConstructor(s_ParamClassName, parent, name, doc);
+        }
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:38]:
```suggestion
        public Param(string parent, string name, string doc)
            : this(SparkEnvironment.JvmBridge.CallConstructor(s_ParamClassName, parent, name, doc)
        {
        }
```

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        
+        /// <summary>
+        /// The description of what the <see cref="Param"/> does and how it works including any
+        /// defaults and the current value
+        /// </summary>
+        /// <returns>A description of how the <see cref="Param"/> works</returns>
+        public string Doc
+        {
+            get => (string)_jvmObject.Invoke("doc");
+        }
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:55]:
```suggestion
        public string Doc => (string)_jvmObject.Invoke("doc");
```

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        
+        /// <summary>
+        /// The name of the <see cref="Param"/>
+        /// </summary>
+        /// <returns></returns>
+        public string Name 
+        {
+            get => (string)_jvmObject.Invoke("name");  
+        }
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:64]:
```suggestion
        public string Name => (string)_jvmObject.Invoke("name");
```

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+        /// <summary>
+        /// The object that contains the <see cref="Param"/>
+        /// </summary>
+        /// <returns></returns>
+        public string Parent
+        {
+            get => (string)_jvmObject.Invoke("parent");
+        }
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:73]:
```suggestion
        public string Parent => (string)_jvmObject.Invoke("parent");
```

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Param/ParamTests.cs [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+        }
+
+        [Fact]
+        public void Test()
+        {
+            const string expectedParent = "parent";
+            const string expectedName = "name";
+            const string expectedDoc = "doc";
+            
+            var param = new Param(expectedParent, expectedName, expectedDoc );
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Param/ParamTests.cs:24]:
```suggestion
            var param = new Param(expectedParent, expectedName, expectedDoc);
```

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace Microsoft.Spark.ML.Feature.Param
+{
+    /// <summary>
+    /// A <see cref="Param"/> with self-contained documentation and optionally default value.
+    ///
+    /// A <see cref="Param"/> references an individual parameter that includes documentation, the
+    /// name of the parameter and optionally a default value. Params can either be set using the
+    /// generic <see cref="Param"/> methods or by using explicit methods. For example
+    /// <see cref="Bucketizer"/> has <c>SetHandleInvalid</c> or you can call <c>GetParam("handleInvalid")</c>
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:17]: over 100 chars.

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+        /// The description of what the <see cref="Param"/> does and how it works including any
+        /// defaults and the current value
+        /// </summary>
+        /// <returns>A description of how the <see cref="Param"/> works</returns>
+        public string Doc => (string)_jvmObject.Invoke("doc");
+        
+        /// <summary>
+        /// The name of the <see cref="Param"/>
+        /// </summary>
+        /// <returns></returns>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:56]: Missing comment

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+        public string Name => (string)_jvmObject.Invoke("name");
+
+        /// <summary>
+        /// The object that contains the <see cref="Param"/>
+        /// </summary>
+        /// <returns></returns>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:62]: ditto

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+    /// <see cref="Param"/> and the value you want to use.
+    /// </summary>
+    public class Param : IJvmObjectReferenceProvider
+    {
+        private static readonly string s_ParamClassName = 
+            "org.apache.spark.ml.param.Param";
+        
+        private readonly JvmObjectReference _jvmObject;
+        
+        public Param(Identifiable parent, string name, string doc)
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:28]: Can we add comments for this since this is public?

File: src/csharp/Microsoft.Spark/ML/Param/Param.cs [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+            : this(SparkEnvironment.JvmBridge.CallConstructor(
+                s_ParamClassName, parent.Uid(), name, doc))
+        {
+        }
+        
+        public Param(string parent, string name, string doc)
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Param/Param.cs:34]: ditto

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Param/ParamTests.cs [hunk 22]
```diff
@@ -0,0 +1,31 @@
+using Microsoft.Spark.ML.Feature.Param;
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Param/ParamTests.cs:1]: Can you add the license header?
